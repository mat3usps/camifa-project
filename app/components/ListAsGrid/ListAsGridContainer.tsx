import type { ReactNode } from "react";
import LinkButton from "~/components/button/LinkButton";
import InactiveItemsList from "~/components/InactiveItemsList/InactiveItemsList";
import ListAsGrid from "~/components/ListAsGrid/ListAsGrid";
import type { LinkTo } from "~/types/LinkTo";

export interface IListAsGridContainerProps<T> {
  activeList: T[];
  inactiveList: T[];
  inactiveTitle: string;
  registerFirstItem: {
    buttonLabel: string;
    title: string;
    to: LinkTo;
  };
  renderItem: (item: T) => ReactNode;
}

const ListAsGridContainer = <T,>({
  activeList,
  inactiveList,
  inactiveTitle,
  registerFirstItem,
  renderItem,
}: IListAsGridContainerProps<T>) => {
  return (
    <>
      {activeList.length < 1 && (
        <>
          <p className="font-bold">{registerFirstItem.title}</p>
          <LinkButton to={registerFirstItem.to} size="lg">
            {registerFirstItem.buttonLabel}
          </LinkButton>
        </>
      )}
      <ListAsGrid>{activeList.map(renderItem)}</ListAsGrid>
      <InactiveItemsList count={inactiveList.length} title={inactiveTitle}>
        {inactiveList.map(renderItem)}
      </InactiveItemsList>
    </>
  );
};

export default ListAsGridContainer;
