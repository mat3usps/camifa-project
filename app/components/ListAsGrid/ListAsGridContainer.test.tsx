import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import type { IListAsGridContainerProps } from "./ListAsGridContainer";
import ListAsGridContainer from "./ListAsGridContainer";

const {
  getAllByTestId,
  getByRole,
  getByTestId,
  getByText,
  queryByRole,
  queryByText,
} = screen;

describe("<ListAsGridContainer />", () => {
  it("should match snapshot", () => {
    const { container } = getRenderer();
    expect(container).toMatchSnapshot();
  });

  it.each([
    ["The title", "Button label"],
    ["Another title", "Another button label"],
  ])(
    "should render register first item with title %s",
    (title, buttonLabel) => {
      getRenderer({
        activeList: [],
        registerFirstItem: {
          buttonLabel,
          title,
          to: "/",
        },
      });
      expect(getByText(title)).toBeInTheDocument();
      expect(getByRole("link", { name: buttonLabel })).toBeInTheDocument();
    }
  );

  it("should NOT render register first item when there is an active item", () => {
    getRenderer({
      activeList: [{ id: "1", name: "Item 1" }],
    });
    expect(queryByText("Register")).not.toBeInTheDocument();
    expect(
      queryByRole("link", { name: "Register item" })
    ).not.toBeInTheDocument();
  });

  it("should render ListAsGrid component for active list", () => {
    getRenderer({
      activeList: [{ id: "1", name: "Item 1" }],
      inactiveList: [],
    });
    expect(getByTestId("ListAsGrid")).toBeInTheDocument();
  });

  it("should render ListAsGrid component for both active and inactive lists", () => {
    getRenderer({
      activeList: [{ id: "1", name: "Item 1" }],
      inactiveList: [{ id: "2", name: "Item 2" }],
    });
    expect(getAllByTestId("ListAsGrid")).toHaveLength(2);
  });

  it("should call renderItem for each item", () => {
    const activeList = [
      { id: "1", name: "Item 1" },
      { id: "2", name: "Item 2" },
    ];
    const inactiveList = [{ id: "3", name: "Inactive Item 1" }];

    const renderItem = vitest.fn();
    getRenderer({
      activeList,
      inactiveList,
      renderItem,
    });
    expect(renderItem).toHaveBeenCalledTimes(
      activeList.length + inactiveList.length
    );
  });

  it("should render all items", () => {
    const activeList = [
      { id: "1", name: "Item 1" },
      { id: "2", name: "Item 2" },
    ];
    const inactiveList = [
      { id: "3", name: "Inactive Item 1" },
      { id: "4", name: "Inactive Item 2" },
    ];

    getRenderer({
      activeList,
      inactiveList,
      renderItem: (item) => <p>{item.name}</p>,
    });
    [...activeList, ...inactiveList].forEach(({ name }) =>
      expect(getByText(name)).toBeInTheDocument()
    );
  });

  it.each(["Inactive items", "Another inactive title"])(
    "should render inactive title %s",
    (expected) => {
      getRenderer({
        inactiveList: [{ id: "3", name: "Inactive Item 1" }],
        inactiveTitle: expected,
      });
      expect(getByRole("heading", { name: expected })).toBeInTheDocument();
    }
  );

  it("should NOT render inactive title", () => {
    getRenderer({
      inactiveList: [],
      inactiveTitle: "Inactive items",
    });
    expect(
      queryByRole("heading", { name: "Inactive items" })
    ).not.toBeInTheDocument();
  });
});

// Helpers
type TestItem = {
  id: string;
  name: string;
};

function getRenderer({
  activeList = [],
  inactiveList = [],
  inactiveTitle = "No items",
  registerFirstItem = {
    buttonLabel: "Register",
    title: "Register item",
    to: "/register",
  },
  renderItem = vitest.fn(),
}: Partial<IListAsGridContainerProps<TestItem>> = {}) {
  return render(
    <MemoryRouter>
      <ListAsGridContainer<TestItem>
        activeList={activeList}
        inactiveTitle={inactiveTitle}
        inactiveList={inactiveList}
        registerFirstItem={registerFirstItem}
        renderItem={renderItem}
      />
    </MemoryRouter>
  );
}
