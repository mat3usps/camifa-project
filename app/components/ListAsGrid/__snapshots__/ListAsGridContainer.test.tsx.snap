// Vitest Snapshot v1

exports[`<ListAsGridContainer /> > should match snapshot 1`] = `
<div>
  <p
    class="font-bold"
  >
    Register item
  </p>
  <a
    class="btn btn-primary btn-lg"
    href="/register"
  >
    Register
  </a>
  <div
    class="grid grid-cols-1 gap-4 sm:grid-cols-2 md:grid-cols-4 lg:grid-cols-6"
    data-testid="ListAsGrid"
  />
</div>
`;
